### PR TITLE
ceph: skip reconcile on PDBs are created, deleted and when on different namespace

### LINKED ISSUE #7011: Nonexistent ceph clusters reported after upgrading from 1.4.7 to 1.5.5
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
After upgrading to rook 1.5.5, I see the following errors show up in the rook operator logs:

```
2021-01-19 21:09:52.722549 E | clusterdisruption-controller: cephcluster "kafka/" seems to be deleted, not requeuing until triggered again
2021-01-19 21:09:52.767254 E | clusterdisruption-controller: cephcluster "kafka/" seems to be deleted, not requeuing until triggered again
```

We have a separate `kafka` namespace that should be totally isolated from rook/ceph.  There is only one Ceph cluster resource in the cluster as reported by kubectl : 

```
 k get cephcluster -A
NAMESPACE   NAME        DATADIRHOSTPATH   MONCOUNT   AGE    PHASE   MESSAGE                        HEALTH
rook-ceph   rook-ceph   /var/lib/rook     3          560d   Ready   Cluster created successfully   HEALTH_OK
```

**Expected behavior:**
No unexpected errors arise in the rook operator log.

**How to reproduce it (minimal and precise):**

Upgraded from Rook 1.4.7 to 1.5.5 and these errors started to appear.

**File(s) to submit**:

* Cluster CR (custom resource), typically called `cluster.yaml`, if necessary
* Operator's logs, if necessary
* Crashing pod(s) logs, if necessary

 To get logs, use `kubectl -n <namespace> logs <pod name>`
When pasting logs, always surround them with backticks or use the `insert code` button from the Github UI.
Read [Github documentation if you need help](https://help.github.com/en/articles/creating-and-highlighting-code-blocks).

**Environment**:
* OS (e.g. from /etc/os-release): Centos 7
* Kernel (e.g. `uname -a`): 3.10.0-1127.19.1.el7.x86_64
* Cloud provider or hardware configuration: baremetal
* Rook version (use `rook version` inside of a Rook Pod): 1.5.5
* Storage backend version (e.g. for ceph do `ceph -v`): v14.2.14-20201119
* Kubernetes version (use `kubectl version`): 1.18.10
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): baremetal
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/master/ceph-toolbox.html)): HEALTH_OK

### PR DESCRIPTION
**Pull request by sp98**
PDB watcher watches for all namespaces. This PR updates the PDB watcher to watch for only operator namespace.
It also updates the watcher not to reconcile for create and delete events.

Signed-off-by: Santosh Pillai <sapillai@redhat.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #7011

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: pkg/operator/ceph/disruption/clusterdisruption/add.go [hunk 1]
```diff
@@ -88,13 +88,22 @@ func Add(mgr manager.Manager, context *controllerconfig.Context) error {
...
+		},
 		UpdateFunc: func(e event.UpdateEvent) bool {
 			pdb, ok := e.ObjectNew.DeepCopyObject().(*policyv1beta1.PodDisruptionBudget)
 			if !ok {
 				return false
 			}
-			// only reconcile if allowed disruptions is 0 in the main PDB
-			return pdb.Name == osdPDBAppName && pdb.Status.DisruptionsAllowed == 0
+			// only reconcile if allowed disruptions for the main PDB is 0 in the operator namespace
+			return e.ObjectNew.GetNamespace() == context.OperatorNamespace &&
```
Comment (travisn) [inline pkg/operator/ceph/disruption/clusterdisruption/add.go:101]: What if we just don't check for the namespace? Then we can still check all namespaces in the scenario where the cluster is in a different namespace.

File: pkg/operator/ceph/disruption/clusterdisruption/add.go [hunk 2]
```diff
@@ -88,13 +88,22 @@ func Add(mgr manager.Manager, context *controllerconfig.Context) error {
 	}
 
 	pdbPredicate := predicate.Funcs{
```
Comment (travisn) [inline pkg/operator/ceph/disruption/clusterdisruption/add.go:90]: Remind me why we need to watch the PDBs? Doesn't the operator act on other events already?

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/disruption/clusterdisruption/add.go (hunk 2 above)
Comment (sp98) [reply pkg/operator/ceph/disruption/clusterdisruption/add.go:90]: After removing the node drain canary,  we needed something that would tell that operator that OSD is down ( or drained) and it should reconcile. This was not possible with the events (other than PDB) that the reconcile acts upon. 

-  Earlier, we were dependent on the cephCluster health check status updating every 60 seconds.  This was not intentional but it did the job. But it was not a good way as user can change the healthCheck duration.  So we decided to make the operator reconcile more proactive. 
- One suggested option was to reconcile every 30 seconds irrespective of any changes. 
- Another approach (currently implemented) was to watch for the Main OSD PDB. With one OSD down (or drained) , the AllowedDisruption of this Main PDB will be 0.  This is what we use to trigger the reconcile now. More in the PR description [here](https://github.com/rook/rook/pull/6749)

File: pkg/operator/ceph/disruption/clusterdisruption/add.go (hunk 1 above)
Comment (leseb) [reply pkg/operator/ceph/disruption/clusterdisruption/add.go:101]: I believe that if we don't, we would reconcile on any PDB updates, right @sp98, checking the ns addresses that issue IIUC.

Comment (sp98) [reply pkg/operator/ceph/disruption/clusterdisruption/add.go:101]: Yes, that's right. 

But one more thing is that,  previously, `CreateFunc` and `DeleteFunc` were not no-op.  Now I've added them as no no-op. So only `pdb.Name == osdPDBAppName && pdb.Status.DisruptionsAllowed == 0` should be enough to ensure that we are not reconciling for PDB updates on other ns (unless some other ns has a pdb named `osdPDBAppName` which is not likely). 

 I feel we can still get rid of `e.ObjectNew.GetNamespace() == context.OperatorNamespace` check.

Comment (travisn) [reply pkg/operator/ceph/disruption/clusterdisruption/add.go:101]: Agreed, seems like the pdb name is a sufficient filter so we can still check all namespaces

File: pkg/operator/ceph/disruption/clusterdisruption/add.go (hunk 2 above)
Comment (travisn) [reply pkg/operator/ceph/disruption/clusterdisruption/add.go:90]: Thanks for the reminder. How about adding a comment here next to the predicate for future reference so I don't ask again? :)

Comment (sp98) [reply pkg/operator/ceph/disruption/clusterdisruption/add.go:92]: added comments.

File: pkg/operator/ceph/disruption/clusterdisruption/add.go (hunk 1 above)
Comment (sp98) [reply pkg/operator/ceph/disruption/clusterdisruption/add.go]: removed namespace check.

**Review by travisn — APPROVED** [review]

**Pull request merged**